### ISSUE TITLE
azurerm_role_definition WaitForState pending state not set correctly

### ISSUE THREAD

**Issue by tiwood** [terraform-providers/terraform-provider-azurerm#10602]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.14.6
+ provider registry.terraform.io/hashicorp/azurerm v2.36.0

### Affected Resource(s)

* `azurerm_role_definition`

### Terraform Configuration Files

```hcl
resource "azurerm_role_definition" "example" {
  name        = "my-custom-role"
  scope       = "ROOT MANAGEMENT GROUP"
  description = "This is a custom role created via Terraform"

  permissions {
    actions     = ["*"]
    not_actions = []
  }

  assignable_scopes = [
    "ROOT MANAGEMENT GROUP"
  ]
}
```

### Debug Output

```
Error: Provider produced inconsistent result after apply

When applying changes to
azurerm_role_definition.custom_roles["workload_contributor.yml"], provider
"registry.terraform.io/hashicorp/azurerm" produced an unexpected new value:
Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

It seems the `WaitForState()` function is called with the wrong `Pending` state property, as the API return `404` if the role is not found YET.

Currently:
```go
	if !d.IsNewResource() {
		id, err := parse.RoleDefinitionId(d.Id())
		if err != nil {
			return err
		}
		stateConf := &resource.StateChangeConf{
			Pending: []string{
				"Pending",
			},
			Target: []string{
				"OK",
			},
			Refresh:                   roleDefinitionUpdateStateRefreshFunc(ctx, client, id.ResourceID),
			MinTimeout:                10 * time.Second,
			ContinuousTargetOccurence: 6,
			Timeout:                   d.Timeout(schema.TimeoutUpdate),
		}

		if _, err := stateConf.WaitForState(); err != nil {
			return fmt.Errorf("waiting for update to Role Definition %q to finish replicating", name)
		}
	}
```

Should be:
```go
	if !d.IsNewResource() {
		id, err := parse.RoleDefinitionId(d.Id())
		if err != nil {
			return err
		}
		stateConf := &resource.StateChangeConf{
			Pending: []string{
				"NotFound",
			},
			Target: []string{
				"OK",
			},
			Refresh:                   roleDefinitionUpdateStateRefreshFunc(ctx, client, id.ResourceID),
			MinTimeout:                10 * time.Second,
			ContinuousTargetOccurence: 6,
			Timeout:                   d.Timeout(schema.TimeoutUpdate),
		}

		if _, err := stateConf.WaitForState(); err != nil {
			return fmt.Errorf("waiting for update to Role Definition %q to finish replicating", name)
		}
	}
```

**Comment by tiwood** [comment]
Ups - its actually only waiting for state if the resource is not new. I'm going to make a PR to also wait if the resource is new, because that's why its failed for me

**Comment by tombuildsstuff** [comment]
@tiwood we've noticed a new role definition gets created behind the scenes during both Create and Update and is eventually reconciled on the backend (it's why #9850 is stalled, since I've not had time to dig into it yet, but that instead switches to checking the CreatedAt/UpdatedAt fields) - it could well be this is related actually 🤔

**Comment by grounded042** [comment]
This is breaking automation for a lot of people. Personally I have Terraform which should be able to automatically run on a daily basis that I now have to manually intervene with every time. Can we get some transparency on a roadmap or timeline on when this issue will be addressed? Coming over from #10442 it seems like this issue is not being properly recognized as a massively disruptive part of many peoples work days.

**Comment by krowlandson** [comment]
Commenting to follow after #10442 was closed

**Comment by JameySteinmann** [comment]
Commenting to follow.

I am having similar and inconsistent issues with creating azurerm_role_definitions. I've tried from 2.44 up to 2.48 with the same inconsistent result of errors

Initially it's a coin toss if the initial role creation complete successfully or throws the error in this issue. After that is the next problem which once I can get terraform to successfully create the role definition and return a proper result, It takes many runs for Terraform to finally be able to assign that role without the below error occurring. 

This is after a 2nd, 3rd or 4th run upon successfully creating the role definition.
```
module.pipelines.azurerm_role_assignment.ado_rotfunctionapp_deployer["rotation"]: Creating...
module.pipelines.azurerm_role_assignment.ado_sasfunctionapp_deployer["sas"]: Creating...

Error: Error loading Role Definition List: could not find role 'My Custom Deployer - dev'

  on modules\function_app_pipelines\Role-Assignments.tf line 9, in resource "azurerm_role_assignment" "ado_rotfunctionapp_deployer":
   9: resource "azurerm_role_assignment" "ado_rotfunctionapp_deployer" {
```

Here is my custom role code along with an assignment if that helps

```
#Role for Pipelines to be able to deploy to FunctionApps
resource "azurerm_role_definition" "pipeline" {
  name = "My Custom Deployer - ${terraform.workspace}"
  scope = "/subscriptions/${var.subscription_id}"
  description = ""
  permissions {
    actions = [
      "Microsoft.Web/sites/Read",
      "Microsoft.Web/sites/config/list/Action", #get publishing creds
      "Microsoft.Web/sites/config/Read",
      "Microsoft.Web/sites/config/write",
      "Microsoft.Web/sites/slots/Read",
      "Microsoft.Web/sites/slots/config/Read"     
    ]
    data_actions = []
    not_actions = []
    not_data_actions = []
  }
  assignable_scopes = [
    "/subscriptions/${var.subscription_id}",
  ]
}

#Assign Pipelines Deployer Role to Service Principal
resource "azurerm_role_assignment" "ado_rotfunctionapp_deployer" {
  for_each = var.rotation_assignment_scopes
  scope = each.value.scope
  role_definition_name = azurerm_role_definition.pipeline.name
  principal_id = azuread_service_principal.pipeline.object_id
  depends_on = [
    azurerm_role_definition.pipeline
  ]
}
```

**Comment by hbuckle** [comment]
The strange thing is we've never had any issues creating role definitions in the past, but in the last week it has started failing consistently. Did anything change in the provider?

**Comment by BrandonE** [comment]
> The strange thing is we've never had any issues creating role definitions in the past, but in the last week it has started failing consistently. Did anything change in the provider?

Same here. Our best workaround is to run it, let it fail, and then import the newly created role definitions into state. After that, everything operates as normal.

**Comment by angelbarrera92** [comment]
> > The strange thing is we've never had any issues creating role definitions in the past, but in the last week it has started failing consistently. Did anything change in the provider?
> 
> Same here. Our best workaround is to run it, let it fail, and then import the newly created role definitions into state. After that, everything operates as normal.

It worked also for me... dealing with Azure terraform provider is.... unstable

**Comment by cosminstirbu** [comment]
@BrandonE  & @angelbarrera92 

How did you find out the role Azure Resource ID that needs to be used with terraform import?

**Comment by BrandonE** [comment]
> @BrandonE & @angelbarrera92
> 
> How did you find out the role Azure Resource ID that needs to be used with terraform import?

`az role definition list --custom-role-only`

**Comment by cosminstirbu** [comment]
@BrandonE - I'm receiving the following error which leads me to think that I'm missing something

`terraform import azurerm_role_definition.rd /subscriptions/.../providers/Microsoft.Authorization/roleDefinitions/...`

`Error: Error parsing Resource ID "/subscriptions/.../providers/Microsoft.Authorization/roleDefinitions/...": could not parse Role Definition ID, invalid format "/subscriptions/.../providers/Microsoft.Authorization/roleDefinitions/..."`

**Comment by andydkelly-ig** [comment]
@cosminstirbu did you see the import section in the [docs] - (https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/role_definition) it's a custom ID that you need to construct yourself. Note the pipe in the middle and the whole ID in quotes.

**Comment by BrandonE** [comment]
> @BrandonE - I'm receiving the following error which leads me to think that I'm missing something
> 
> `terraform import azurerm_role_definition.rd /subscriptions/.../providers/Microsoft.Authorization/roleDefinitions/...`
> 
> `Error: Error parsing Resource ID "/subscriptions/.../providers/Microsoft.Authorization/roleDefinitions/...": could not parse Role Definition ID, invalid format "/subscriptions/.../providers/Microsoft.Authorization/roleDefinitions/..."`

I have the resource in a submodule. This is how I'm importing it into state while fetching my Subscription ID from the Azure CLI and parsing it with jq:

`terraform import module.<Sub Module>.azurerm_role_definition.<Role Definition Label> "/subscriptions/$(az account list | jq -r '.[0].id')/providers/Microsoft.Authorization/roleDefinitions/<Role Definition ID>|/subscriptions/$(az account list | jq -r '.[0].id')"`

**Comment by cosminstirbu** [comment]
Thank you both. After reading the docs (and also figuring out that my scope is not `/subscriptions/00000000-0000-0000-0000-000000000000`, but rather `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/my-prod-rg/providers/Microsoft.Network/virtualNetworks/my-prod-vnet`) I managed to implement the import workaround.

**Comment by mybayern1974** [comment]
This issue is expected to have been fixed in the service backend, which does not require Terraform changes. People who hit this issue please have a try.

**Comment by tullydwyer** [comment]
Hi @mybayern1974 ,

I was having this issue before and have just re-ran my IAC. Terraform successfully created 17/18 of the azurerm_role_definition's and got the inconsistent result error for the last one. It created the role in Azure, but is not tracked in Terraform. Repeat runs say that the role already exists. Ran in Azure DevOps.

TF Version: 0.14.7
azurerm provider: 2.49.0

See logs below.

First run:
```txt
Error: Provider produced inconsistent result after apply

When applying changes to
module.REDACTED.azurerm_role_definition.REDACTED,
provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new
value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

Releasing state lock. This may take a few moments...
```

Second run:
```txt
Error: A resource with the ID "/subscriptions/***/providers/Microsoft.Authorization/roleDefinitions/REDACTED|/subscriptions/***" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_role_definition" for more information.

  on ../../../REDACTED.tf line 1, in resource "azurerm_role_definition" "REDACTED":
   1: resource "azurerm_role_definition" "REDACTED" {


Releasing state lock. This may take a few moments...
```

Kind regards,

**Comment by lvisterin** [comment]
I personally haven't run into this issue since @mybayern1974 's comment anymore, but we don't have as many role definitions as the poster above.

**Comment by andydkelly-ig** [comment]
I have been able to deploy our infra for the first time today since this issue began 👍

**Comment by BrandonE** [comment]
> This issue is expected to have been fixed in the service backend, which does not require Terraform changes. People who hit this issue please have a try.

Do you have documentation on the fix? This issue was catastrophic for me and I want to ensure it never happens again.

**Comment by mybayern1974** [comment]
Hi all, I posted my findings here after seeing things became stable in my local Terraform project involving the role definition story, with comparing with previous symptom that I got 50% failure rate. With saying that I'm sorry but I do not have context on the backend. 

Hi BrandonE, I can have a try to possibly contact corresponding backend support to see whether there could be justification provided if that could help.

**Comment by BrandonE** [comment]
@mybayern1974 That would be greatly appreciated and can hopefully lead to the closure of this ticket!

**Comment by alastairtree** [comment]
I am another customer whose deploys have all started working reliably again for a few days suggesting this has been fixed.

**Comment by BrandonE** [comment]
This is working for me now, but I'd really appreciate official word from Azure to understand why this happened and how we can know that this won't happen in the future.

**Comment by varunkch** [comment]
In order to improve call latencies we had introduced distributed caching of Role Definitions. As a result of this change some customers experienced failures when a role definition is read immediately after writing. The root cause of the failure was that the read request was served from cache which was not updated immediately after the write. We have disabled the distributed caching of Role Definitions which has fixed the failures.

The root cause has been well understood and we will take extreme care in the future to prevent repeat occurrences of similar failures for this scenario and other scenarios. 

There is still a small probability that read after write scenarios can experience this failure since we cache Role Definitions in-memory as well. The in-memory cache is not new and we have had it for a long time. So the probability of this issue happening again is the same as it was before we added the distributed cache.

Thank you all for the feedback which helps make our services better.
